src/ResultWithoutOption.d.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2016 Tetsuharu OHZEKI <saneyuki.snyk@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+type Maybe<T> = T | void;
+
+type MapFn<T, U> = (v: T) => U;
+type FlatmapOkFn<T, U, E> = (v: T) => Result<U, E>;
+type FlatmapErrFn<T, E, F> = (e: E) => Result<T, F>;
+type RecoveryFn<E, T> = (e: E) => T;
+
+/**
+ *  The Result/Either type interface whose APIs are inspired
+ *  by Rust's `std::result::Result<T, E>`.
+ */
+export type Result<T, E> = Ok<T, E> | Err<T, E>;
+
+interface ResultMethods<T, E> {
+    /**
+     *  Returns true if the result is `Ok`.
+     */
+    isOk(): this is Ok<T, E>;
+
+    /**
+     *  Returns true if the result is `Err`.
+     */
+    isErr(): this is Err<T, E>;
+
+    /**
+     *  Converts from `Result<T, E>` to `T | void`.
+     *  If the self is `Ok`, returns `T`.
+     *  Otherwise, returns `void`.
+     */
+    ok(): Maybe<T>;
```

Comment:
*__Design Question__*:

If this type is specialized to `Result<void, ErrorMessage>`, this method would always returns `void` (`undefined`). Is this really no problem?

In some case, we'd like to express the return type which has "some error massages" and "empty success value". Then, we should not use `Option<ErrorMessage>` because this means "if you have some errors, then it returns `Some<E>`. Otherwise (in success/valid case), it returns `None`". This is NOT expressive and confuse user. So we should have some usecases to use `Result<void, ErrorMessage>`. How should we treat this case?